src/site/_includes/components/Prerender.js

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const {html} = require('common-tags');
+
+/**
+ * Prerender by adding Speculation Rules JSON
+ * Also adds a GA event so can see how many prerender attempts were made
+ * versus prerender navigations.
+ * @param {string} url The url to prerender
+ * @returns {string}
+ */
+module.exports = (url) => {
+  const ORIGIN = 'https://web.dev';
```

Comment:
Actually, rather than defining this here, you should probably import this from the site data. e.g.

```js
const site = require('../../_data/site');

// ...

  if (!url || new URL(url, site.url).origin !== site.url) {
    return '';
  }
```

Reply:
Last thought: rather than returning `''`, maybe we should throw an error instead. e.g.:

```js
const {html, oneLine} = require('common-tags');
const site = require('../../_data/site');

// ...

  if (!url || new URL(url, site.url).origin !== site.url) {
    throw new Error(oneLine`
      Cannot create prerender rule for URL: "${url}",
      only same-origin prerendering is supported at this time.
    `);
  }
```

I think throwing a build error is better than silently failing (where it may not be obvious why the shortcode didn't work).

Reply:
Done, but I wonder if this will ever cause problems? For example, we currently prerender the first /blog/ article and next course page. What if they are ever external links? In that case the build will fail. I don't _think_ they ever should be, but if we roll this out more, we may start to see that so one to keep an eye on.